### PR TITLE
Introduce `content_type` to `azurerm_storage_blob`

### PR DESCRIPTION
**Pull request by JunyiYi**
* user can specify the blob MIME type (Azure accepts all kinds of string not only MIME, so I didn't add validation rule here)
* the MIME type will be returned in HTTP header when user access the output `url` of the blob
* the MIME type is updatable and unlike other properties in SDK, it is a plain string rather than a pointer

This resolves issue #999 .

### CODE REVIEW

File: azurerm/resource_arm_storage_blob.go [hunk 1]
```diff
@@ -149,6 +156,7 @@ func resourceArmStorageBlobCreate(d *schema.ResourceData, meta interface{}) erro
 	blobType := d.Get("type").(string)
 	cont := d.Get("storage_container_name").(string)
 	sourceUri := d.Get("source_uri").(string)
+	contentType := d.Get("content_type").(string)
 
 	log.Printf("[INFO] Creating blob %q in storage account %q", name, storageAccountName)
 	if sourceUri != "" {
```
Comment (metacpp) [inline azurerm/resource_arm_storage_blob.go]: Refer to this article to simplify the `happy-path` code block:
https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

Comment (tombuildsstuff) [reply azurerm/resource_arm_storage_blob.go]: we can pull this out into separate methods - but I think that's outside of the scope of this PR tbh; so let's do that in a separate PR

Comment (JunyiYi) [reply azurerm/resource_arm_storage_blob.go]: Agree with @tombuildsstuff . The blob storage code is a messy and I also think the fields are not good and make people confused. That will be a big refactoring, so let's do it in another PR if possible.

File: azurerm/resource_arm_storage_blob.go [hunk 2]
```diff
@@ -53,6 +54,12 @@ func resourceArmStorageBlob() *schema.Resource {
 				Default:      0,
 				ValidateFunc: validateArmStorageBlobSize,
 			},
+			"content_type": {
+				Type:          schema.TypeString,
```
Comment (jeffreyCline) [inline azurerm/resource_arm_storage_blob.go]: Should do a state migration test here? Just to make sure it doesn't break anything downstream?

Comment (JunyiYi) [reply azurerm/resource_arm_storage_blob.go]: I have run the new `terraform plan` after the old `terraform apply`, and it turns out nothing has been changed. I think there is not needs to do state migration here. @katbyte @tombuildsstuff @paultyng what do you think?

File: azurerm/resource_arm_storage_blob.go [hunk 3]
```diff
@@ -53,6 +54,12 @@ func resourceArmStorageBlob() *schema.Resource {
...
+				Optional:      true,
+				Default:       "application/octet-stream",
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_storage_blob.go]: where does this default come from?

File: azurerm/resource_arm_storage_blob.go (hunk 2 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_storage_blob.go]: related question below: where does this default value come from?

File: azurerm/resource_arm_storage_blob.go (hunk 1 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_storage_blob.go]: also, at _some point in the future_ (read: when the SDK's are more stable) we'll be switching over to the new [Storage SDKs](http://github.com/Azure/azure-storage-blob-go/issues/48); which would probably be a sensible time to do this refactoring

File: azurerm/resource_arm_storage_blob_test.go [hunk 4]
```diff
@@ -666,3 +711,45 @@ resource "azurerm_storage_blob" "destination" {
 }
 `, rInt, location, rString, sourceBlobName)
 }
+
+func testAccAzureRMStorageBlobPage_blockContentType(rInt int, rString, sourceBlobName, location, contentType string) string {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_storage_blob_test.go]: _minor_ we've got a _vague_ convention here of making these `(rInt, rString, location, [others])`

File: azurerm/resource_arm_storage_blob.go (hunk 2 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_storage_blob.go]: can we also document this new field in the docs (`website/docs/r/storage_blob.html.markdown`)

File: azurerm/resource_arm_storage_blob.go [hunk 5]
```diff
@@ -556,6 +600,13 @@ func resourceArmStorageBlobRead(d *schema.ResourceData, meta interface{}) error
 
 	container := blobClient.GetContainerReference(storageContainerName)
 	blob := container.GetBlobReference(name)
+
+	err = blob.GetProperties(nil)
+	if err != nil {
+		return fmt.Errorf("Error getting properties of blob %s (container %s, storage account %s): %+v", name, storageContainerName, storageAccountName, err)
+	}
+	d.Set("content_type", blob.Properties.ContentType)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_storage_blob.go]: can we nil-check if `Properties != nil` here?

File: azurerm/resource_arm_storage_blob.go (hunk 2 above)
Comment (JunyiYi) [reply azurerm/resource_arm_storage_blob.go]: Sure. The default value comes from `Content-Type` section in https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob#request-headers-all-blob-types.

File: azurerm/resource_arm_storage_blob.go (hunk 3 above)
Comment (JunyiYi) [reply azurerm/resource_arm_storage_blob.go]: The default value comes from `Content-Type` section in https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob#request-headers-all-blob-types.

File: azurerm/resource_arm_storage_blob.go [hunk 6]
```diff
@@ -524,6 +535,39 @@ func resourceArmStorageBlobBlockUploadWorker(ctx resourceArmStorageBlobBlockUplo
...
+	storageContainerName := d.Get("storage_container_name").(string)
+
+	container := blobClient.GetContainerReference(storageContainerName)
+	blob := container.GetBlobReference(name)
+
+	if d.HasChange("content_type") {
+		blob.Properties.ContentType = d.Get("content_type").(string)
+	}
+
+	err = blob.SetProperties(nil)
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_storage_blob.go]: The default timeout 30s should be just fine: https://docs.microsoft.com/en-us/rest/api/storageservices/setting-timeouts-for-blob-service-operations.

File: azurerm/resource_arm_storage_blob.go [hunk 7]
```diff
@@ -556,6 +600,13 @@ func resourceArmStorageBlobRead(d *schema.ResourceData, meta interface{}) error
 
 	container := blobClient.GetContainerReference(storageContainerName)
 	blob := container.GetBlobReference(name)
+
+	err = blob.GetProperties(nil)
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_storage_blob.go]: The default timeout 30s should be just fine: https://docs.microsoft.com/en-us/rest/api/storageservices/setting-timeouts-for-blob-service-operations.

File: azurerm/resource_arm_storage_blob.go (hunk 5 above)
Comment (JunyiYi) [reply azurerm/resource_arm_storage_blob.go]: `Properties` is not a pointer type: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/vendor/github.com/Azure/azure-sdk-for-go/storage/blob.go#L34. Actually both `Properties` and `ContentType` are not pointer types.

**Comment by katbyte** [comment]
Tests pass: 
screen shot 2018-06-11 at 23 15 53

**Pull request merged**